File: Desktop/ДЗ Python/DZ4.py
# Задача 26:  Напишите программу, которая на вход принимает два числа A и B,
# и возводит число А в целую степень B с помощью рекурсии.

# def power(x, n):
#     if n == 0:
#         return 1
#     if n < 0:
#         return 1/power(x, -n)
#     if n % 2 == 0:
#         return power(x, n // 2) * power(x, n // 2)
#     else:
#         return power(x, n - 1) * x
#
# x = int(input())
# n = int(input())
#
# print(power(x, n))

# Задача 28: Напишите рекурсивную функцию sum(a, b), возвращающую сумму двух целых неотрицательных чисел.
# Из всех арифметических операций допускаются только +1 и -1. Также нельзя использовать циклы.

# def sum(a, b):
#     if b == 0:
#         return a
#     if a == 0:
#         return b
#     else:
#         return sum(a, b - 1) + 1
# 
# a = int(input())
# b = int(input())
# 
# print(sum(a, b))

File: Desktop/JAVAproject/DZ4.java
// // Task_3

// import java.util.ArrayDeque;
// import java.util.Arrays;
// import java.util.Deque;

// public class Task_3 {

//     public static void main(String[] args) {
//         Deque<Integer> deque = new ArrayDeque<>(Arrays.asList(1, 2, 3, 4, 5, 6));
//         System.out.println(checkOn(deque));
//         deque = new ArrayDeque<>(Arrays.asList(1, 2, 3, 3, 2, 1));
//         System.out.println(checkOn(deque));
//     }

//     public static boolean checkOn(Deque<Integer> deque) {
//         while (deque.size() > 1) {
           
//             if (!deque.removeFirst().equals(deque.removeLast())) {
//                 return false;
//             }
//         }
//         return true;
//     }
// }


// // Task_4


// import java.util.ArrayDeque;
// import java.util.Arrays;
// import java.util.Deque;

// public class Task_4 {

//     public static void main(String[] args) {
//         Deque<Integer> d1 = new ArrayDeque<>(Arrays.asList(1,2,3));
//         Deque<Integer> d2 = new ArrayDeque<>(Arrays.asList(5,4,7));
//         Deque<Integer> result = sum(d1, d2);
//         System.out.println(result); // [6, 6, 0, 1]
//     }

//     public static Deque<Integer> sum(Deque<Integer> d1, Deque<Integer> d2) {
//         Deque<Integer> result = new ArrayDeque<>();

//         int carry = 0; // запомненная единица
//         while (!d1.isEmpty() || !d2.isEmpty() || carry != 0) {
//             int sum = carry;
//             if (!d1.isEmpty()) sum += d1.removeFirst();
//             if (!d2.isEmpty()) sum += d2.removeFirst();
//             if (sum > 9) {
//                 carry = 1;
//                 sum -= 10;
//             } else {
//                 carry = 0;
//             }
//             result.addLast(sum);
//         }

//         return result;
//     }

// }


// Task_5

// public boolean validate(String input) {
//     Deque<Character> stack = new ArrayDeque<Character>();
//     for (int i = 0; i < input.length(); i++) {
//         char c = input.charAt(i);
//         if (c == '(' || c == '{' || c == '[') {
//             stack.push(c);
//         } else if (c == ')' || c == '}' || c == ']') {
//             if (stack.isEmpty()) {
//                 return false;
//             }
//             char prev = stack.pop();
//             if (prev == '(' && c != ')' || prev == '{' && c != '}' || prev == '[' && c != ']') {
//                 return false;
//             }
//         }
//     }
//     return stack.isEmpty();
// }

